linked_list/linked_list.py

```diff
@@ -16,113 +16,186 @@ def __init__(self):
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
     # Time Complexity: ?
     # Space Complexity: ?
     def get_at_index(self, index):
-        pass
+        current_node = self.head
+        if index >= self.length():
+            return None
+        i = 0
+        while current_node != None:
+            if i == index:
+                return current_node.value
+            i += 1
+            current_node = current_node.next
+        return None
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
     # Time Complexity: ?
     # Space Complexity: ?
     def get_last(self):
-        pass
+        current_node = self.head
+        while current_node != None:
+            if current_node.next == None:
+                return current_node.value
+            else: 
+                current_node = current_node.next
+        return None
 
     # method that inserts a given value as a new last node in the linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def add_last(self, value):
-        pass
+        current_node = self.head
+        new_node = Node(value)
+        while current_node != None:
+            if current_node.next == None:
+                current_node.next = new_node
+                break
+            current_node = current_node.next
+        else: 
+            self.head = new_node
+
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head == None:
+            return None
+        max = self.head.value
+        current_node = self.head
+        while current_node != None:
+            if current_node.value > max:
+                max = current_node.value
+            current_node = current_node.next
+        return max
+
 
     # method to delete the first node found with specified value
     # Time Complexity: ?
     # Space Complexity: ?
     def delete(self, value):
-        pass
+        current_node = self.head
+        if not current_node:
+            return None
+        if current_node.value == value:
+            self.head = current_node.next
+            return 
+        while current_node != None:
+            if current_node.next:
+                if current_node.next.value == value:
+                    print(current_node.value, current_node.next.value)
+                    current_node.next = current_node.next.next
+                    return
+            current_node = current_node.next
+            print(current_node.value, current_node.next.value, "last")
+           
+    
 
     # method to print all the values in the linked list
     # Time Complexity: ?
     # Space Complexity: ?
-    def visit(self):
-        helper_list = []
-        current = self.head
+    # def visit(self):

```

Comment:
For `visit` you can:
1.  Create an empty list
1.  Traverse the linked list like you did previously and at each iteration you can append the current value to the list
1.  You can then `return ", ".join(the_list)`